### ISSUE TITLE
[libb2] build failure with Android NDK r26 canary

### ISSUE THREAD

**Issue by equeim** [microsoft/vcpkg#31332]
### Operating system

Linux, cross-compiling for Android

### Compiler

Clang

### Steps to reproduce the behavior

Download NDK r26 canary build from https://ci.android.com/builds/branches/aosp-master-ndk/grid?legacy=1

```Shell
ANDROID_NDK_HOME=/path/to/android-ndk-r26-canary ./vcpkg install libb2:arm-neon-android
```

### Failure logs

[config-arm-neon-android-dbg-config.log](https://github.com/microsoft/vcpkg/files/11424878/config-arm-neon-android-dbg-config.log)
[config-arm-neon-android-dbg-err.log](https://github.com/microsoft/vcpkg/files/11424879/config-arm-neon-android-dbg-err.log)
[config-arm-neon-android-dbg-out.log](https://github.com/microsoft/vcpkg/files/11424880/config-arm-neon-android-dbg-out.log)

### Additional context

configure script fails with:
```
configure:3804: checking whether the C compiler works
configure:3826: /home/alexey/Android/android-ndk-r26-canary/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -g -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -mthumb -Wformat -Werror=format-security -fPIC -fno-limit-debug-info --target=armv7-none-linux-androideabi21 -DANDROID -D_FORTIFY_SOURCE=2 -march=armv7-a --sysroot=/home/alexey/Android/android-ndk-r26-canary/toolchains/llvm/prebuilt/linux-x86_64/sysroot -avoid-version --target=armv7-none-linux-androideabi21 -static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--no-undefined-version -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments --sysroot=/home/alexey/Android/android-ndk-r26-canary/toolchains/llvm/prebuilt/linux-x86_64/sysroot conftest.c -latomic -lm >&5
clang: error: unknown argument: '-avoid-version'
```

This flag is added in https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_configure_make.cmake#L792

BTW NDK r26 will enter beta soon.

**Comment by FrankXie05** [comment]
@equeim 
The `TARGET` is not `ANDROID`.
```
Target: x86_64-unknown-linux-gnu
```

**Comment by equeim** [comment]
That's normal `clang -v` prints android target only if you pass `--target` option (r25 does the same). I think the reason for error is that previously Clang would silently ignore `-avoid-version` option (which is meant for libtool, not for Clang) and now it fails with `error: unknown argument: '-avoid-version'`. This results in configure tests failing because they are invoked without libtool but use LDFLAGS.

**Comment by dg0yt** [comment]
Do we have a port which is know to build shared objects for android?
I failed to build libiconv or gmp due to linker errors listing `-lgcc` with NDK r25c.
In other words, is there any use in carrying around `-avoid-version` at the moment.

**Comment by equeim** [comment]
> Do we have a port which is know to build shared objects for android?
> I failed to build libiconv or gmp due to linker errors listing `-lgcc` with NDK r25c.
> In other words, is there any use in carrying around `-avoid-version` at the moment.

This is another issue which is libtool dropping `--target` option when linking shared libraries (for some reason it doesn't do so when linking executables which is why build doesn't fail for ports containing tools). I tried to fix it in this PR but decided to drop it for now.

`-avoid-version` needs to be fixed anyways because it leads to build failures when `./configure` is running its try-compile tests regardless whether you build static or shared library.

Currently I'm waiting on #31228 because this PR depends on its changes with LDFLAGS handling.

**Comment by Neumann-A** [comment]
> libtool dropping --target option when linking 

does this also happen when `--target` is forced to be in CC/CXX instead ?

**Comment by dg0yt** [comment]
> Currently I'm waiting on #31228 because this PR depends on its changes with LDFLAGS handling.

And I'm trying to understand what I need to pick into #31228 to solve this issue, too. Touching these scripts causes a lot of rebuilds, so I don't want to force it on users too often. And unfortunately, many rebuilds means finding many baseline regressions.

In the current run I just removed `-avoid-version`. We might just patch the generated `libtool` for android. But I don't want to add this without having test ports.

**Comment by dg0yt** [comment]
> > libtool dropping --target option when linking
> 
> does this also happen when `--target` is forced to be in CC/CXX instead ?

Prefixing the target with `-XCClinker` in LDFLAGS works. Now I still need to avoid-version.

**Comment by dg0yt** [comment]
> Now I still need to avoid-version.

No, I don't. The packages just need to use a recent libtool. libiconv already works now (NDK25c). libbacktrace's libtool is tool old.
(New version of libtool set `version_type=none`.)

**Comment by dg0yt** [comment]
> Prefixing the target with `-XCClinker` in LDFLAGS works. Now I still need to avoid-version.

... unless trying with the NDK 26 preview. Well, it is the same problem as with `-avoid-version`: It is a pure libtool option.

> > does this also happen when `--target` is forced to be in CC/CXX instead ?

I see that this is already used for the osx and ios targets. Trying that.

**Comment by equeim** [comment]
> > Prefixing the target with `-XCClinker` in LDFLAGS works. Now I still need to avoid-version.
> 
> ... unless trying with the NDK 26 preview. Well, it is the same problem as with `-avoid-version`: It is a pure libtool option.
> 

I think Clang ignores -Xcompiler option so you can probably prefix with that.

**Comment by dg0yt** [comment]
Not stale.
